cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java

```diff
@@ -707,6 +728,24 @@ private String cleanse(String val, int maxLength) {
     return cleansed.length() > maxLength ? cleansed.substring(0, maxLength) : cleansed;
   }
 
+  /**
+   * Creates per-run configmap containing config files.
+   */
+  private void createConfigMap() throws ApiException, IOException {
+    String configMapName = CONFIGMAP_NAME_PREFIX + twillRunId.getId();
+    V1ConfigMapBuilder builder = new V1ConfigMapBuilder().withMetadata(new V1ObjectMeta().name(configMapName));
+    String runnableName = getMainRuntimeSpecification(twillSpec.getRunnables()).getName();
+    for (LocalFile localFile : twillSpec.getRunnables().get(runnableName).getLocalFiles()) {
+      // Only add xml config files to the configmap. We skip over jars as they are too large store in a
+      // configmap (max size = 1 MB. See https://kubernetes.io/docs/concepts/configuration/configmap).
+      if (!localFile.getName().endsWith(".xml")) {
```

Comment:
anything that gets localized through the configmap does not need to be localized through the file-localizer. How does the file-localizer know what it needs to pull? Should make sure we're not pulling things twice in different ways.

Reply:
For user programs, xml config files are pulled from per-run configmap, all other files (jars, archives, program options, appspec) are pulled by file-localizer.

We don't have a per-run configmap for system programs. file-localizer is responsible for pulling all the files in this case.

Reply:
Localfiles are added in the twillspec in `populateRunnableLocalFiles()`. twillspec gets deserialized in file-localizer